### ISSUE TITLE
getAllChildrenAsOrderedTree returning some child terms in the wrong level

### ISSUE THREAD

**Issue by ohermans1** [pnp/pnpjs#2448]
### Category
- [ ] Enhancement
- [x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [ 3.8.0 ]

Please specify what version(s) of SharePoint you are targeting: [ Sharepoint Online ]

### Expected / Desired Behavior / Question
Related to issue [getAllChildrenAsOrderedTree not returning copied/reused terms](https://github.com/pnp/pnpjs/issues/2414)

Following the recent update to getAllChildrenAsOrderedTree  I would expect the fuction to return an ordered tree, something is going wrong, which is meaning some child terms are being returned at the wrong level. 

### Observed Behavior
Hi team! I have updated to v3.8 and have run a little testing on my end and have run across some potentially odd behavior.

First of all, for some context, for this test I am creating an ordered list using the following:
![image](https://user-images.githubusercontent.com/50157976/196816476-08f9fb66-f004-4fb0-a566-1e388dbb5f7e.png)

Which is pointed to the 'Process Zone' termset
![image](https://user-images.githubusercontent.com/50157976/196816528-36fcd887-6ec3-4bdc-86f9-659d4759a34f.png)

What I'm finding when I run this code and print it to the console is that 'Child 1' and 'Child 2' are appearing at the top level, rather than under 'Term 3', but this doesn't seem to occur for 'Term 1.1' and 'Term 1.2' (Which sit as expected under 'Term 1')
![image](https://user-images.githubusercontent.com/50157976/196816565-57594fd3-8f42-4bf5-a901-33a1fd178e89.png)

Any ideas?

### Steps to Reproduce
Update to 3.8.0, run getAllChildrenAsOrderedTree on a termset as per my screenshot above.

**Comment by patrick-rodgers** [comment]
Hi @ohermans1, I am not able to reproduce this as described. I setup a termset with three terms and the child terms as you show. I get back the correct tree no matter how I call the method (with or without retrieveProperties).

Do you know in what order you created the terms in your set? Were there created and reorderd? Anything else you can share?

**Comment by ohermans1** [comment]
Hi @patrick-rodgers - Thanks for checking this out for me. 

I've had a good look and can't determine anything odd about these particular terms! So, I'm really not sure. They may have been created and reordered/moved at some point (do you think that could cause an issue), but aside from that they are all just regular terms under my Global Term Groups.

**Comment by patrick-rodgers** [comment]
@ohermans1 - hmmm, not sure. I don't know what is causing this so was just asking, I don't think moving/reordering should matter but 🤷‍♂️. If you could describe a way to create a term set that fails I can investigate further, until I can reproduce the issue I'd just be guessing.

**Comment by ohermans1** [comment]
Hi @patrick-rodgers - Sorry again for the massive gap between my responses! Got tied up in something else for a while. 

I've just been taking a look back at this - First of all, I've updated to 3.9. Then I have recreated a completely fresh site level term set for testing. 

I am getting the same outcome (the tree not quite being created properly), with a completely fresh, unmodified termset. 

Are you able to check if you get the same?
![image](https://user-images.githubusercontent.com/50157976/201767195-ad20eb28-e64e-4ab9-8227-6504f30f4616.png)
![image](https://user-images.githubusercontent.com/50157976/201767376-55b701b2-56cc-4186-87cd-e487ab1253c1.png)

Here is the exact code I am running to get the ordered tree (with SharePointSetup being spfi().using(SPFx(this.context));)
```
const getTerms = async () => {
    const sp = props.SharePointSetup;
    const store = new PnPClientStorage();
    //getAllChildrenAsOrderedTree is expensive, so cached for 30 mintues
    const childTreeCached: IOrderedTermInfo[] = await store.local.getOrPut(
      "myKey",
      () => {
        return sp.termStore.groups
          .getById(props.processZoneTermId)
          .sets.getById(props.processZoneTermId)
          .getAllChildrenAsOrderedTree({ retrieveProperties: true });
      },
      dateAdd(new Date(), "minute", 30)
    );
    console.log(
      "🚀 ~ file: ProcessZonesSelector.tsx ~ line 53 ~ getTerms ~ childTreeCached",
      childTreeCached
    );

    return childTreeCached;
  };
```

**Comment by juliemturner** [comment]
Do you have this problem if you take caching out?

**Comment by ohermans1** [comment]
Hi Julie - Sorry, I saw your message and thought I had responded - but obviously not!

Yep, I still get the exact same issue if I remove caching. I've tried setting up completely fresh termsets and everything and the same issue occurs every time. My colleagues and I can't see anything obvious I might be doing wrong on my end - so its a tricky one! Is it working fine for you and the team if you do a similar setup?

**Comment by juliemturner** [comment]
Yes, as @patrick-rodgers said above we cannot replicate this issue. And unfortunately, unless we can reproduce it, we cannot fix it. I would encourage you to do some debugging of your own and see if you can give us more information, if we can figure out how to reproduce it and/or if you can tell us what's wrong you could submit a PR and/or give us more information so we could help.

**Comment by ohermans1** [comment]
Hey @juliemturner, good idea, if it's not recreateable from your end, it could well be something I've done, so I will investigate further and let you know if I find anything interesting. Thanks so much for your quick responses and for taking the time to take a look. Ill reach out with more info if I come across anything. :)

**Comment by ohermans1** [comment]
Hey @juliemturner and @patrick-rodgers - Sorry to jump back onto this one, but I have been trouble shooting over the last day or so, and have found something interesting which may help you recreate?

First of all, I created a completely fresh simple termset as follows 
![image](https://user-images.githubusercontent.com/50157976/204158279-0efdc545-e8eb-4892-9ce1-0fe3c315ff34.png)

Next, I created a fresh webpart and ran the following code to get my ordered tree. 

```
sp = spfi().using(SPFx(this.context));

aysnc const getTerms = () => {
    const childTree: IOrderedTermInfo[] = await this.sp.termStore.sets
      .getById(theTermSetID)
      .getAllChildrenAsOrderedTree({
        retrieveProperties: true,
      });
    console.log("🚀 ~ getTerms ~ childTree", childTree);
  }
```
As a result, I got a correctly ordered tree with 3 children (as per the term store layout!). Woo!
![image](https://user-images.githubusercontent.com/50157976/204158577-1d5b9c31-03ef-4ad3-b444-1d7b4842701c.png)

I then continued to experiment a little further and found some odd behavior. If I renamed one of the child terms so that its label was alphabetically lower than its parent, I once again go the weird behaviout where the ordered tree wasnt set out correctly. In my above example for instance, if I rename Term1.3 to Child1.3, it will now no longer appear under 'Term1' on the ordered list and will instead appear at the same level as 'Term1'. You can also see from my screenshot below, that 'Term1' now only has the 2 children in its children array. 
![image](https://user-images.githubusercontent.com/50157976/204158761-40e25ffb-76ab-4a74-b002-ce6f8263b5fe.png)

I did some tests with other term labels and had the same results - if the child label is alphabetically below the parent, it appears out of order. If I rename the terms back to their original setup the tree once again behaviors as expected. 

Sorry about the long message - but I just thought this was quite an interesting discovery - does this make any sense from your end?

**Comment by patrick-rodgers** [comment]
Interesting, with that I am able to recreate the issue so I'll have a look. Thanks for continuing to dig in.

**Comment by ohermans1** [comment]
Hey @patrick-rodgers - I have installed the latest nightly version and can confirm the function behaving as expected! 

I also just wanted to say thanks to yourself and @juliemturner - you've both been an awesome help and the pnpjs is great!

Cheers again

Ollie

**Comment by juliemturner** [comment]
Thanks so much, really glad it's fixed!

**Comment by patrick-rodgers** [comment]
Thanks @ohermans1 - appreciate the kind words, it means a lot. Glad the library is useful for you!
